lib/Email/MIME/CreateHTML.pm

```diff
@@ -127,19 +129,70 @@ sub parts_for_objects {
 	return @html_mime_parts;
 }
 
-sub build_html_email {
-	my($header, $html, $body_attributes, $html_mime_parts, $plain_text_mime) = @_;
+# ( $string, $status ) = _normalize_to_perl_string( $string, $encoding )
+#
+# Tries (no guarantee) to ensure that the string passed in becomes a
+# decoded perl unicode string; i.e. not an encoded sequence of octets.
+#
+# In the optimal case the given string already is a decoded perl unicode string
+# (or ascii), in which case it simply returns it with an undef status.
+#
+# If not, it tries to make it a decoded perl unicode string, and
+# returns the status as well to explain what it thinks the string was.
+
+sub _normalize_to_perl_string {
+	my ($string, $encoding) = @_;
+	if(ref(my $enc = guess_encoding $string)) { # tests ascii and utf8.
+		# Need to accept UTF8 without even checking what the target encoding was,
+		# as the input may correctly be a decoded perl unicode string, but
+		# the target for the email may be some other encoding.
+		my $backup = $string;
+		_utf8_on $string; # NOP on decoded perl unicode or ascii strings.
```

Comment:
Isn't there the possibility that this could result in an internally corrupt perl string?  If perl sets the utf8 bit then it will assume that the string has a valid utf8 encoding, and you end up returning this string....

what about replacing this with
```
utf8::decode($string) unless utf8::is_utf8($string)
```
(I didn't test this, but the idea is to have perl verify that it decodes from utf8 bytes, but don't do that if perl marked this as already having happened)

Reply:
Guess::Encoding only returns on valid encodings, so this should be fine.